### ISSUE TITLE
Support recursive TypedDict fields when using the TypedDict functional syntax

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#4151]
### Discussed in https://github.com/microsoft/pyright/discussions/4150

<div type='discussions-op-text'>

<sup>Originally posted by **jawshooah** November  6, 2022</sup>
Getting errors in strict mode for the following code. Pyright only seems to struggle with the recursive `foo` field on `FooVar`, which is defined using the `TypedDict` functional syntax. It has no issues with `FooCls`, which is defined using the `TypedDict` class syntax.

```python
# pyright: strict
from typing import NotRequired, TypedDict


class FooCls(TypedDict):
    foo: NotRequired["FooCls"]
    bar: NotRequired[str]


FooVar = TypedDict("FooVar", {"foo": NotRequired["FooVar"], "bar": NotRequired[str]})


foo_cls: FooCls = {"foo": {"bar": "bar"}}
foo_var: FooVar = {"foo": {"bar": "bar"}}


def get_bar_cls(foo: FooCls) -> str:
    if "bar" in foo:
        return foo["bar"]
    if "foo" in foo:
        return get_bar_cls(foo["foo"])
    raise ValueError(foo)


def get_bar_var(foo: FooVar) -> str:
    if "bar" in foo:
        return foo["bar"]
    if "foo" in foo:
        return get_bar_var(foo["foo"])
    raise ValueError(foo)
```

Pyright output:
```
❯ pyenv exec pyright typed_dict_error.py
No configuration file found.
No pyproject.toml file found.
stubPath /private/tmp/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
pyright 1.1.278
/private/tmp/typed_dict_error.py
  /private/tmp/typed_dict_error.py:14:19 - error: Expression of type "dict[str, FooVar]" cannot be assigned to declared type "FooVar"
    "bar" is required in "FooVar" (reportGeneralTypeIssues)
  /private/tmp/typed_dict_error.py:29:28 - error: Argument type is unknown
    Argument corresponds to parameter "foo" in function "get_bar_var" (reportUnknownArgumentType)
2 errors, 0 warnings, 0 informations 
Completed in 0.548sec
```</div>

**Comment by erictraut** [comment]
This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.279, which I just published. It will also be included in a future version of pylance.
